### PR TITLE
C++: extensible range analysis

### PR DESCRIPTION
**Pull request by jbj**
This PR integrates @lcartey's extensible range analysis prototype into `SimpleRangeAnalysis`, adding tests and documentation as well as some fixes based on problems I encountered along the way.

The commit history is a bit more "back and forth" than what I usually do. That's because I wanted to preserve @lcartey's authorship on the original three commits.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll [hunk 1]
```diff
@@ -1504,3 +1520,25 @@ private module SimpleRangeAnalysisCached {
...
   }
 }
+
+/**
+ * INTERNAL: do not use. This module contains utilities for use in the
+ * experimental `SimpleRangeAnalysisExpr` module.
+ */
+module SimpleRangeAnalysisInternal {
+  /**
+   * Gets the truncated lower bounds of the fully converted expression.
```
Comment (intrigus-lgtm) [inline cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:1530]: In which way are these bounds truncated?

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:1530]: Truncated to fit in the range of the type. For example, the truncated upper bound of an `unsigned char` will never be more than 255.

File: cpp/ql/src/experimental/semmle/code/cpp/models/interfaces/SimpleRangeAnalysisExpr.qll [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  /**
+   * Holds if this expression depends on the range of its unconverted
+   * subexpression `child`. This information is used to inform the range
+   * analysis about cyclic dependencies. Without this information, range
+   * analysis might work for simple cases but will go into infinite loops on
+   * complex code.
+   *
+   * For example, when modeling a function call whose return value depends on
+   * all of its arguments, implement this predicate as
+   * `child = this.getAnArgument()`.
```
Comment (geoffw0) [inline cpp/ql/src/experimental/semmle/code/cpp/models/interfaces/SimpleRangeAnalysisExpr.qll:53]: Good explanation. :+1:

**Comment by jbj** [comment]
> How's performance? I'm hopeful it has no effect when there are no custom overrides of `SimpleRangeAnalysisExpr` in scope?

I tested performance on a MySQL snapshot during development and found no slowdown, but I lost the exact numbers.

Now I've tested on Wireshark and found no slowdown either (with no override of `SimpleRangeAnalysisExpr` in scope apart from `Empty`). For the whole `SimpleRangeAnalysis` stage, the time column in `CSV_IMB_QUERIES` changes from 199.174 to 199.986, and the number of rows remains unchanged at 18744015. The logs don't mention `SimpleRangeAnalysisExpr` at all, so it must have been compiled out.

**Pull request merged**